### ISSUE TITLE
Megolm Backup | Client should always have a local copy of all their megolm keys to limit the possibility of key loss

### ISSUE THREAD

**Issue by BillCarsonFr** [vector-im/element-meta#2446]
We do not proactively download message keys from key storage (and instead, download keys on demand after a UTD).

This means that, if the user then resets their key backup (for example because they forgot their recovery key), it is likely that a significant number of message keys will be lost.

To address this, we will make each new client download (in the background) a copy of all the message keys in key storage as soon as it can. The operation should be resilient to app restart and should have a UI to report progress.

We'd rather not implement this twice, so our plan as of this writing is to switch Element Web to using the key-backup management code in matrix-rust-sdk, and then implement this once there.

Tasks:
 * [ ] https://github.com/matrix-org/matrix-rust-sdk/issues/3611
 * [ ] rust-sdk: Ensure that all the keys from a given backup version are downloaded
 * [ ] EIX: UI feedback for ongoing import
 * [ ] EAX: UI feedback for ongoing import
 * [ ] https://github.com/element-hq/element-web/issues/27009

<details><summary>Previous description</summary>

## Context

We implemented a recent change to the way [we get the room keys from backup](https://github.com/element-hq/element-meta/issues/757). Previously, the full backup was downloaded after verification, as soon as the backup decryption key was known, but that did not result in a good user experience when your backup was too big. The time to download the full backup and decrypt it could be taking 10s of minutes depending on the platform.

In order to improve that (and in the absence of [pagination](https://github.com/element-hq/element-meta/issues/2176) for the backup), keys are now downloaded on-demand when we try to decrypt an historical message.
This results in a better user experience, as the messages get decrypted faster after a new login.

## Problem

With this new way of getting the keys, all user sessions will now have only a partial set of the backed up keys. This can cause key loss if for example a user tries to reset his backup (without importing first from the previous backup).

Notice that the previous implementation of full backup import has several issues:
- [Performance](https://github.com/element-hq/element-web/issues/23359)
- Lack of [pause/stop/resume](https://github.com/element-hq/element-web/issues/17950). Particularly on mobile, if the app was put in background and terminated by the OS before the end of import, it was not restarted.
- Was possible to start several import in parallel

## Solution

We want to ensure that each new client gets a copy of all the keys in backup as soon as it can. The operation should be resilient to app restart and should have an API [richvdh: or, more to the point: a UI?] to report progress.

* EAX/EIX:
  - [ ] rust-sdk: Ensure that all the keys from a given backup version are downloaded
  - [ ] EIX: UI feedback for ongoing import
  - [ ] EAX: UI feedback for ongoing import
- Web (https://github.com/element-hq/element-web/issues/27009):
  - [ ] js-sdk: Ensure that all the keys from a given backup version are downloaded
  - [ ] https://github.com/element-hq/element-web/issues/17839
- Element Android:
- Element iOS

## Related/Existing issues
- https://github.com/element-hq/element-web/issues/27009
- https://github.com/element-hq/element-web/issues/20694

</details>

**Comment by richvdh** [comment]
I wrote some notes on how we might do this for web at https://github.com/element-hq/element-web/issues/27009.

Notes:
 * In a naive implementation, if the app is killed during the recovery, we'll restart from the beginning next time. Anything else is somewhat annoying without a fix to https://github.com/element-hq/element-meta/issues/2176.
 * If your backup is too big, it's possible that we'll never successfully download it; the client will just keep retrying.

**Comment by richvdh** [comment]
Question: is there any value in attempting to download individual keys on-demand once the full backup has been successfully downloaded? I'd think not?

**Comment by andybalaam** [comment]
@richvdh will remodel this item based on https://docs.google.com/document/d/1eEj_BagixDuny_fhF_WjoKEm0SD538mTXA3s436-c2I/edit - we chose the "Update Rust backup manager, and refactor EW to use it" option

**Comment by BillCarsonFr** [comment]
> Question: is there any value in attempting to download individual keys on-demand once the full backup has been successfully downloaded? I'd think not?

Short answer: Once it has been successfully downloaded, no. But while it's being downloaded definitely yes.
If not it will take ages before I can see history after a login (unless we have [this](https://github.com/element-hq/element-meta/issues/2176)).

Longer answer:
- Maybe a quick race where when you start downloading, some new keys are being uploaded and you'd want to get them after?
- Maybe in case of identity reset some devices might become unsigned for some time and then be excluded from key distribution, but when they are verified again they could re-try the backup?

**Comment by andybalaam** [comment]
Matthew raised this with me today. Right now if a user loses their recovery key and all their verified devices so their only option is to reset their identity, we don't have a way of making sure they have all the keys locally before they do that, so they risk losing history.

If we complete this issue, resetting identity does not need to be so scary, because the user will be able to check that they have all keys downloaded locally, meaning the reset will not lose any history. In fact, our UI could become less scary if we detect that all keys are downloaded.

**Comment by ara4n** [comment]
Another option could be to *not* proactively store a local copy of all your keys, but instead have the "Export keys" button first downloads all keys from the server (not unreasonable)?

**Comment by andybalaam** [comment]
> Another option could be to _not_ proactively store a local copy of all your keys, but instead have the "Export keys" button first downloads all keys from the server (not unreasonable)?

Yes, this would be a good stopgap, and less work. I still want the full feature in the longer run though - it proactively prevents all kinds of problems, including if your homeserver goes away.

**Comment by BillCarsonFr** [comment]
> user loses their recovery key and all their verified devices so their only option is to reset their identity, we don't have a way of making sure they have all the keys locally before they do that, so they risk losing history.

I am not sure to understand, in that case it is too late, nothing could help? all your keys are lost.

The current issue is more this scenario:
- Your first device is EW1. You have it for very long. All the keys are cached locally.
- After sometime you login a new Mobile session EX1, you verify it.

At this point EW1 has all histotical keys **H_full**
But EX1 only have partial historical keys **H_partial**

Now you change computer and you log out EW1.
You only have EX1 with **H_partial**

And if you want to login a new web EW2, and you don't have access to EX1, then you will reset recovery on the new login.
One would think that it is fine, I'll later verify EX1 and it will re-upload the keys. But that would only re-upload **H_partial**.

Isn't **instead have the "Export keys" button first downloads all keys from the server** jusk asking to add back the `Import all keys from backup` button? 
And expect user to do that after a new login and wait for it to complete? and ensure the phone doesnt dose out

**Comment by andybalaam** [comment]
Our focus at the moment is on preventing situations requiring people to reset their identity from occurring. We'd love to get to this when we have time, but we are prioritising prevention for the moment.
